Enable complex-valued matrices and vectors

BarbaraV opened the pull request:
I tried to implement complex-valued matrices and vectors, please have a look at it. Some quick (and easy) tests worked for me. Still, there are three points I'm not quite sure about:

1)whether the implementation of IstlRowMajorSparseMatrix.valid() really detects nan and infinity. Maybe, the two_norm() rather gives an error when an entry is nan or infinity -- I haven't tested this yet.
2) whether it is a good idea to implement float_cmp::eq also for complex-valued xtype and ytype. However, this is needed for complex-valued vectors in the member function almost_equal().
3) the static assertion for ScalarImp=double in EigenMappedDenseVector: I left it as it was, but now it gives a failure in the method valid() for complex-valued EigenRowMajorSparseMatrix.

ftalbrecht commented:
Thank you very much for your work! Concerning your questions:

* I am also not sure if `valid` does the right thing for the `IstlRowMajorSparseMatrix` either. If you are unsure, you could implement it the same way you did for the `CommonDenseMatrix`, i.e., checking the real and imaginary part of each entry seperately.
* I also think that we should handle `float_cmp` a bit different: I think you only have to adapt the (first) implementation in line 95: https://github.com/BarbaraV/dune-stuff/blob/master/dune/stuff/common/float_cmp.hh#L95. If you implement this one twice (once for complex, once for arithmetic types), then you do not have to change any of the other methods below (I think). The way you could do that would be as follows:
We would first need a way to statically decide if something is a `std::complex` (to be used in `std::enable_if`) by implementing `is_complex`. Therefore, add the following code at the top of `float_cmp.hh` (line 17):
```c++
namespace internal {


template< class T >
struct is_complex_helper
{
  DSC_has_typedef_initialize_once(value_type)

  static const bool is_candidate = DSC_has_typedef(value_type)< T >::value;
}; // class is_complex_helper


} // namespace internal


template< class T, bool candidate = internal::is_complex_helper< T >::is_candidate >
struct is_complex
  : public std::is_base_of< std::complex< typename T::value_type >, T >
{};


template< class T >
struct is_complex< T, false >
  : public std::false_type
{};

```
Using this, you can leave the implementation of `float_cmp` in line 95 as it is and add directly below a variant
```c++
template< class T >
    typename std::enable_if< is_complex< T >::value, bool >::type
float_cmp(const T& xx, const T& yy, const typename T::value_type& rtol, const typename T::value_type& atol)
{
...
}
```
where inside, `xx` and `yy` are guaranteed to be a `std::complex` and you can extract the real and imaginary part and just call the other `float_cmp` for each of them.
This way you do not have to change any other implementation in that file and we get a `float_cmp` for `std::complex` for free, which is also nice.
* The third one is bit tricky. You are right to leave the assertion in the vector, that one is only implemented for `double`. This does, however, pose a problem for the matrix. I would suggest to comment out the implementation of `valid` in the `EigenRowMajorSparseMatrix` completely. Then, the default implementation from the interface should take over. That one is significantly slower but should do the job. You should then add a new issue like _valid() of EigenRowMajorSparseMatrix not optimal any more_ and we can deal with this problem later.

I would also like you to change the following, before merging:
* Concerning the naming I think that in the long run we should conform to the choice in dune-common: we should call what is now the `ScalarType` `FieldType` and what you called `RealScalarType` just `RealType`. The change of `ScalarType` to `FieldType` might be a bit tricky and can be left for future work but I would like you to change `RealScalarType` to `RealType`.
* I would also like you to do some whitespace changes. I am aware that I failed to document our coding policy until now so I am sorry for the additional work. But I will highlight one or two places in the code directly that should give you an idea.

Once all of this is done I am more than happy to merge, being able to support `std::complex` is very nice. In the long run we should think about adding corresponding tests (perhaps @tobiasleibner could do that since the original tests where his, but this can wait). In addition, I suspect similar problems for the linear solvers, but we can tackle those later, too.

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -36,6 +37,7 @@ struct VectorAbstraction
   typedef VecType VectorType;
   typedef VecType ScalarType;
```
You should add
```c++
typedef VecType RealType
```
here.

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -58,8 +60,11 @@ template< class T >
+  typedef T ScalarType;
   typedef ScalarType       S;
+  //typedef typename Dune::FieldTraits< T >::real_type RealScalarType;
+  //typedef typename RealScalarType R;
```
Do you have a particular reason not to use the variant you commented out here?

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -82,8 +87,10 @@ template< class K >
   typedef ScalarType               S;
+  typedef typename Dune::FieldTraits< K >::real_type RealScalarType;
+  typedef RealScalarType R;
 
```
This is one example for indentation. Could you please change these lines to 
```c++
typedef Dune::DynamicVector< K >                    VectorType;
typedef typename Dune::FieldTraits< K >::field_type ScalarType;
typedef typename Dune::FieldTraits< K >::real_type  RealType;
typedef ScalarType S;
typedef RealType   R;
```
so that the indentation of the defined types lines up.

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -106,8 +113,10 @@ template< class K, int SIZE >
+  typedef typename Dune::FieldTraits< K >::field_type ScalarType;
   typedef ScalarType                   S;
+  typedef typename Dune::FieldTraits< K >::real_type RealScalarType;
+  typedef RealScalarType R;
```
see above

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -136,8 +145,10 @@ template< class K, int SIZE >
+  typedef typename Dune::FieldTraits< K >::field_type ScalarType;
   typedef ScalarType                                  S;
+  typedef typename Dune::FieldTraits< K >::real_type RealScalarType;
+  typedef RealScalarType R;
```
see above

ftalbrecht commented on dune/stuff/la/container/common.hh:
```diff
@@ -45,7 +47,8 @@ template< class ScalarImp = double >
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef CommonDenseVector< ScalarType >   derived_type;
   typedef Dune::DynamicVector< ScalarType > BackendType;
 };
```
see above

ftalbrecht commented on dune/stuff/la/container/common.hh:
```diff
@@ -55,7 +58,8 @@ template< class ScalarImp = double >
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef CommonDenseMatrix< ScalarType >   derived_type;
   typedef Dune::DynamicMatrix< ScalarType > BackendType;
 };
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/base.hh:
```diff
@@ -57,6 +59,7 @@ class EigenBaseVector
+  typedef typename Traits::RealScalarType RealScalarType;
   typedef typename Traits::BackendType  BackendType;
   typedef typename Traits::derived_type VectorImpType;
 
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/dense.hh:
```diff
@@ -60,7 +62,8 @@ template< class ScalarImp = double >
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef EigenDenseVector< ScalarType > derived_type;
   typedef typename ::Eigen::Matrix< ScalarType, ::Eigen::Dynamic, 1 > BackendType;
 }; // class EigenDenseVectorTraits
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/dense.hh:
```diff
@@ -74,7 +77,8 @@ class EigenMappedDenseVectorTraits
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef EigenMappedDenseVector< ScalarType > derived_type;
   typedef Eigen::Map< PlainBackendType >       BackendType;
 }; // class EigenMappedDenseVectorTraits
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/dense.hh:
```diff
@@ -87,7 +91,8 @@ template< class ScalarImp = double >
-  typedef ScalarImp                      ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::field_type ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef EigenDenseMatrix< ScalarType > derived_type;
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/dense.hh:
```diff
@@ -101,17 +106,18 @@ public:
   typedef EigenDenseVector< ScalarImp >                                               ThisType;
   typedef VectorInterface< internal::EigenDenseVectorTraits< ScalarImp >, ScalarImp > VectorInterfaceType;
-  typedef EigenBaseVector< internal::EigenDenseVectorTraits< ScalarImp > >            BaseType;
+  typedef EigenBaseVector< internal::EigenDenseVectorTraits< ScalarImp >, ScalarImp >            BaseType;
```
see above

ftalbrecht commented on dune/stuff/la/container/eigen/dense.hh:
```diff
@@ -206,11 +212,11 @@ private:
   typedef EigenMappedDenseVector< ScalarImp >                                               ThisType;
   typedef VectorInterface< internal::EigenMappedDenseVectorTraits< ScalarImp >, ScalarImp > VectorInterfaceType;
-  typedef EigenBaseVector< internal::EigenMappedDenseVectorTraits< ScalarImp > >            BaseType;
+  typedef EigenBaseVector< internal::EigenMappedDenseVectorTraits< ScalarImp >, ScalarImp >            BaseType;
```
s.a.

ftalbrecht commented on dune/stuff/la/container/eigen/sparse.hh:
```diff
@@ -55,7 +57,8 @@ template< class ScalarImp = double >
+  typedef typename Dune::FieldTraits< ScalarImp >::field_type ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef EigenRowMajorSparseMatrix< ScalarType > derived_type;
   typedef typename ::Eigen::SparseMatrix< ScalarType, ::Eigen::RowMajor > BackendType;
```
s.a.

ftalbrecht commented on dune/stuff/la/container/istl.hh:
```diff
@@ -51,7 +53,8 @@ template< class ScalarImp >
+  typedef typename Dune::FieldTraits< ScalarImp >::field_type ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef IstlDenseVector< ScalarImp >                derived_type;
   typedef BlockVector< FieldVector< ScalarType, 1 > > BackendType;
```
s.a.

ftalbrecht commented on dune/stuff/la/container/istl.hh:
```diff
@@ -64,7 +67,8 @@ template< class ScalarImp >
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
   typedef IstlRowMajorSparseMatrix< ScalarType >        derived_type;
   typedef BCRSMatrix< FieldMatrix< ScalarType, 1, 1 > > BackendType;
 }; // class RowMajorSparseMatrixTraits
```
s.a.

ftalbrecht commented on dune/stuff/la/container/matrix-interface.hh:
```diff
@@ -45,7 +47,8 @@ class MatrixInterface
-  typedef ScalarImp                     ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::field_type ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
 
```
s.a.

ftalbrecht commented on dune/stuff/la/container/vector-interface.hh:
```diff
@@ -48,7 +51,8 @@ class VectorInterface
-  typedef ScalarImp ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::field_type ScalarType;
+  typedef typename Dune::FieldTraits< ScalarImp >::real_type RealScalarType;
 
```
s.a.

ftalbrecht commented on dune/stuff/la/container/vector-interface.hh:
```diff
@@ -664,6 +669,8 @@ struct VectorAbstractionBase
   typedef ScalarType S;
+  typedef typename std::conditional< is_vector, typename VectorImp::RealScalarType, void >::type RealScalarType;
+  typedef RealScalarType R;
 
```
s.a.

BarbaraV commented:
Thanks for reviewing and the suggestions for improvement. I will incorporate your suggestions soon.

I also noticed that most likely the solvers have to be adapted, too. 

Concerning valid() in IstlRowMajorSparseMatrix: It cannot be implemented in the same way as the in the CommonDenseMatrix because the return type of entry is a DuneFieldMatrix<std::complex, 1, 1> and therefore the compiler complains about using std:abs or std::real. If we are sure that the block sizes are always 1x1, one can of course handle it by checking entry[0][0].

BarbaraV commented:
I implemented your suggestions, thank you for your kind help! I hope the use of whitespaces corresponds to the coding policy now. Thanks for remarking it, it will try to remember it in the future.

 In two (small) points, I didn't follow your suggestions:
1) One cannot comment out valid() in EigenRowMajorSparseMatrix as the interface just forwards the method to the derived classes and by removing valid(), one gets an CHECK_CRTP failure. I liked more to get the static_assertion failure as I think it is more understandable. Nevertheless, I'll open up a corresponding issue in an instant.

2) in float_cmp.hh Iused your nice is_complex_helper, but renamed its template parameter Tt. Otherwise I got the (quite mysterious) compiler error:

 dune/stuff/common/float_cmp.hh:24:3: error: declaration of 'T' shadows template parameter
  DSC_has_typedef_initialize_once(value_type)
  ^
dune/stuff/common/type_utils.hh:159:22: note: expanded from macro 'DSC_has_typedef_initialize_once'
  template< typename T > \
                     ^
dune/stuff/common/float_cmp.hh:21:17: note: template parameter is declared here
template< class T >

Now it works for me and, I hope, in the correct way.

ftalbrecht commented on dune/stuff/common/vector.hh:
```diff
@@ -135,9 +144,11 @@ struct VectorAbstraction< Dune::FieldVector< K, SIZE > >
-  typedef ScalarType                                  S;
+  typedef Dune::Stuff::Common::FieldVector< K, SIZE >  VectorType;
+  typedef typename Dune::FieldTraits< K >::field_type  ScalarType;
+    typedef typename Dune::FieldTraits< K >::real_type RealType;
```
Fix indentation here.

ftalbrecht commented:
Ah, I think I used `T` in the macro, which is not the best choice perhaps. Thanks for spotting this, I will fix it myself! I left you one more comment on indentation. Additionally: in `float_cmp.hh`, you changed all `std::is_same< ... ::S, ... >` to`std::is_same< ... ::R, ... >`. I think it should rather remain `S`, instead of `R` (since `S` is the full type, e.g., `std::complex< double >` while `R` would be only `double`, correct?). Since you added the `float_cmp` specialization that will work for `S` and `R` we can compare the full type everywhere and not worry about `R`, since the new specialization will take care of that (and has a static check for that).
If you agree, please change the `R` to `S` again in those checks. Since I am currently on the road I will merge this in the next days and trigger the automated build to see if everything will work.

BarbaraV commented:
Thanks for spotting the indentation mistake!

With your remarks on float_cmp.hh, I do not really agree. If I change R back to S and I want to compare two complex vectors (e.g. in almost_equal of the vector-interface), the compiler will not find an implementation of float_cmp as the std::is_same between VectorType::S (aka complex) and ToleranceType (aka double) fails. The new float_cmp specialization I added is only called for the components (in l. 147). So I think it is right that we check the RealType of the vector and the ToleranceType for equality in l. 138/139. 

I changed back the R into S for the dune_float_cmp, though, as this is only implemented for non-complex types.

ftalbrecht commented:
You are right, I agree. I though the `std::is_same` were comparing the field types of the vectors, but they are just comparing each to the tolerance, which should be real. I will merge your changes soon...

renemilk commented:
@BarbaraV  you should possibly glance over the recent changes I did wrt std::complex usage, that should make some code simpler to write. I hope I didn't break anything existing. If I did please do add an appropiate failing test.
